### PR TITLE
2197 websocket connector

### LINKED ISSUE #2197: Websocket augur-sdk connector
Ads a websocket promise style connector.

### PR DESCRIPTION
**Pull request by brianosaurus**
Creates a websocket connector with a related test.

Needs event handling and ID tracking.

Closes #2197

### CODE REVIEW

File: packages/augur-sdk/src/connector/ws-connector.test.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { WebsocketConnector } from "./ws-connector";
```
Comment (justinbarry) [inline packages/augur-sdk/src/connector/ws-connector.test.ts:1]: This isn't a proper jest test. I think @robert-e-davidson3 has thoughts on how to test this.

File: packages/augur-sdk/src/connector/ws-connector.ts [hunk 2]
```diff
@@ -1,25 +1,42 @@
...
+  }
+
+  public async disconnect(): Promise<any> {
+    return this.socket.close();
   }
 
   public bindTo<R, P>(f: (db: any, augur: any, params: P) => R): (params: P) => Promise<R> {
-    return async (params: P): Promise<R> => {
-      return {} as R;
+    return async (params: P): Promise<any> => {
```
Comment (justinbarry) [inline packages/augur-sdk/src/connector/ws-connector.ts]:
```suggestion
    return async (params: P): Promise<R> => {
```

File: packages/augur-sdk/src/constants.ts [hunk 3]
```diff
@@ -2,16 +2,56 @@ import { utils as ethersUtils } from "ethers";
...
+  U_PORT = "uPort",
+  LEDGER = "ledger",
+  PRIVATE_KEY = "privateKey",
+  UNLOCKED_ETHEREUM_NODE = "unlockedEthereumNode",
+  META_MASK = "metaMask",
+  TREZOR = "trezor",
+  EDGE = "edge",
+}
+
+export enum SubscriptionEventNames {
```
Comment (justinbarry) [inline packages/augur-sdk/src/constants.ts:14]:
```suggestion
export enum SUBSCRIPTION_EVENT_NAMES {
```
We have been using uppercase underscore for enums.

File: packages/augur-sdk/src/constants.ts [hunk 4]
```diff
@@ -2,16 +2,56 @@ import { utils as ethersUtils } from "ethers";
...
+  TokensTransferred = "TokensTransferred",
+  TradingProceedsClaimed = "TradingProceedsClaimed",
+  UniverseCreated = "UniverseCreated",
+  Burn = "Burn",
+  TokensBurned = "TokensBurned",
+  Approval = "Approval",
+}
 
-export const ETHER = new ethersUtils.BigNumber(10).pow(18)
+export enum ControlMessageType {
```
Comment (justinbarry) [inline packages/augur-sdk/src/constants.ts:43]:
```suggestion
export enum CONTROL_MESSAGE_TYPE {
```

File: packages/augur-sdk/src/events.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { EventEmitter } from "events";
+import { SubscriptionEventNames } from "./constants";
+
+// Some events, like MarketState, are not always sourced from logs and do not have an "eventName"
+// Always make sure it is present without having to specify in every .emit call
+class EventNameEmitter extends EventEmitter {
+  public emit(type: SubscriptionEventNames, ...args: Array<any>): boolean {
```
Comment (justinbarry) [inline packages/augur-sdk/src/events.ts:7]:
```suggestion
  public emit(type: SUBSCRIPTION_EVENT_NAMES, ...args: Array<any>): boolean {
```

File: packages/augur-sdk/src/events.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { EventEmitter } from "events";
+import { SubscriptionEventNames } from "./constants";
```
Comment (justinbarry) [inline packages/augur-sdk/src/events.ts:2]:
```suggestion
import { SUBSCRIPTION_EVENT_NAMES } from "./constants";
```

File: packages/augur-sdk/src/state/WebsocketEndpoint.ts [hunk 7]
```diff
@@ -6,10 +6,14 @@ import WebSocket from "ws";
 
 import { API } from "./api/API";
 import { AddressFormatReviver } from "./AddressFormatReviver";
-import { JsonRpcRequest, EndpointSettings } from "./api/types";
+import { ControlMessageType } from "../constants";
```
Comment (justinbarry) [inline packages/augur-sdk/src/state/WebsocketEndpoint.ts:9]:
```suggestion
import { CONTROL_MESSAGE_TYPE } from "../constants";
```

File: packages/augur-sdk/src/state/WebsocketEndpoint.ts [hunk 8]
```diff
@@ -52,12 +56,17 @@ export async function run(api: API, endpointSettings: EndpointSettings): Promise
   });
 
   servers.push(new WebSocket.Server({ server }));
+  controlEmitter.emit(ControlMessageType.ServerStart);
```
Comment (justinbarry) [inline packages/augur-sdk/src/state/WebsocketEndpoint.ts:59]:
```suggestion
  controlEmitter.emit(CONTROL_MESSAGE_TYPE.ServerStart);
```

**Review by justinbarry — CHANGES_REQUESTED** [review]
Made some variable name suggestions.

File: packages/augur-sdk/src/connector/ws-connector.ts (hunk 2 above)
Comment (brianosaurus) [reply packages/augur-sdk/src/connector/ws-connector.ts]: done

File: packages/augur-sdk/src/connector/ws-connector.test.ts (hunk 1 above)
Comment (brianosaurus) [reply packages/augur-sdk/src/connector/ws-connector.test.ts:1]: 👍

**Review by justinbarry — APPROVED** [review]

**Pull request merged**